### DISCUSSION TITLE
Suggested way to pass files from the client

### DISCUSSION THREAD

**Discussion by nickwinder** [modelcontextprotocol/modelcontextprotocol#1197] [Q&A - General]
### Pre-submission Checklist

- [X] I have checked that this question would not be more appropriate as an issue in a specific repository
- [X] I have searched existing discussions and documentation for answers

### Question Category

- [X] Protocol Specification
- [ ] SDK Usage
- [ ] Server Implementation
- [ ] General Implementation
- [ ] Documentation
- [ ] Other

### Your Question

👋
I’m exploring ways to enable LLMs to process files using MCP servers. As a first step, the client needs to pass or report the file to the server for processing.

From the specification, I see two potential approaches:
1.	Ask the LLM to send a base64 blob as part of the tool call.
    - This approach is unreliable, particularly as file sizes increase.
2.	Request the [roots/list](https://spec.modelcontextprotocol.io/specification/client/roots/#listing-roots) endpoint from the client to identify readable files in the file system.
    - This depends on the client correctly implementing the roots protocol. However, in my testing with version 0.7.1 of the Claude desktop app, the roots protocol doesn’t appear to be supported (requests fail with -32601 (Method not found)).

Does anyone have suggestions for better ways to pass files from the client to the server for processing? Am I overlooking something fundamental in the protocol?

**Comment by jerome3o-anthropic** [comment]
Hey @nickwinder - thanks for reaching out, I'm on the anthropic team and this is something that I've been looking into recently, and want to figure out a good way to implement this sort of thing at the protocol level. I don't think you're overlooking anything, your use-case is currently finicky in the current state of the protocol.

My initial thoughts/work have been about allowing MCP server tools to return "files" - basically a tool result content type that is not intended for the LLM's context window, but more for use in an conversation's workspace. For an example scenario, this is what I'd like to enable:

* A BigQuery (or other SQL/data warehouse tool) is used by an LLM to make an SQL query that returns too much data for the LLM context (say something in the order of \~10mb)
  * This is returned from the tool using a tool result content "file" type
  * The "file" tool result indicates to the MCP host that this shouldn't enter the prompt verbatim, but it's creation should be indicated to the LLM (i.e. by adding "file XYZ was created")
* The LLM uses another tool to process that file
  * The processing tool then saves a file, which is also returned to the MCP host as a tool result of the "file" type
* The LLM then loads this file back to the warehouse, or elsewhere, using the processed file

This effectively allows bespoke ETL style workflows to be done by the LLM using MCP (along with many other use cases).

My thinking so far is to extend the [tool result format](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.json#L88) to have a `FileContent` type, with B64 encoded data as the content. This would solve the "Tools creating files" usecase, but I'm not sure how we could implement the intention for the model to provide a file to another MCP server - I'm open to ideas.

Some things that are top of mind here for me are:

* We don't want to make too many assumptions around this "meta-context" should work for other apps, i.e. do we expect most MCP hosts to maintain a filesystem?
* How should the model indicate that it wants to include a file, and how should we accommodate this in the protocol

**Reply by nickwinder** [reply]
Thanks for replying! 
I follow what you're saying, but one thing I may suggest changing is the way "file" data is reported and recieved. 
For the [tool result](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.json#L88) you refer to, I feel it would make more sense to have a new `URIResource` which can refer to a URI, but importantly does not contain the content blob. 
The client is then free to fetch that content blob from the Server if it feels the need to do so. 
I see some concepts in the schema that already relate to this kind of behavior, namely [ReadResourceResult](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.json#L1412), and [ReferenceResource](https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.json#L1559C10-L1559C27). These seem like they can be borrowed to make what you describe possible fairly easily. 

As for the client to provide a file (or resource) to a server, I would imagine, a similar mechanism must be reported to the server in order to process. 
- Model sends a tool request with a URI
- Server can query client for the given URI and retrieve a blob to work with

I believe I need to sit with the protocol for longer to fully understand its limitations, but this seems logical to me right now.

**Comment by evalstate** [comment]
I started a related discussion in the Specification area here: https://github.com/modelcontextprotocol/specification/discussions/90

I think this is what the "Audience" annotation is for - 
https://github.com/modelcontextprotocol/specification/blob/bb5fdd282a4d0793822a569f573ebc36804d38f8/schema/schema.ts#L814-L836 

But this doesn't quite answer how best to deal with the specific implementation (which could be Roots, Resource Templates, or custom handling). 

At the moment, the Client doesn't expose Resource Templates to the server; having a symmetrical approach might make sense for this case where the client can offer Metadata and Resource Identifiers that are made available to the server (and potentially scoped to a session). 

This feels cleaner than a shared file system via Roots. 

Another thing I'm finding is that I'm enjoying playing with MCP Servers and having Claude orchestrate them - but in a lot of cases once I've seen 2 clients working together, I don't necessarily need inference between them - some way of chaining them might be good.

**Comment by cpryland** [comment]
@evalstate Any progress here?

**Comment by mheubi** [comment]
I ran across the same challenge when  implementing a PDF text extractor tool as an MCP tool.
Since these PDFs can be arbitrarily large, I do not want the extracted text to be loaded directly into the LLM context, but rather stored into a text file which can then be pulled in piecewise as the LLM sees fit.

While this can all be done by having an external file location and passing file references to and from the tool, it is much more preferable from a security perspective to pass files to and from the tool directly.

This is where I arrived at the same approach as @jerome3o-anthropic is suggesting, introducing a file return type:
 
**result = [types.FileContent(type="file", data=data, name='myresult.txt (optional)', mimeType='text/markdown)]**

The client then stores that file locally (in my case in a virtual file system inside the client's session). It then only loads the VFS file-id into LLM context. The LLM can now access the file or split it into chunks (depending on size) using additional client side tools.

P.S: When initiating the tool call, I upload the file to the tool based on annotations in the tool description, which indicate which tool parameters are file-ids. My generic MCP client, upon detcting a tool invocation, first POSTs all the referenced files from its VFS to the MCP server (which has its own VFS, aka  a session dictionary ;-). This post operation lives outside the MCP spec, so  for my proprietary use case, no real protocol change is required in the upstream direction. Having an official upstream mechanism as well, however, would allow for cross-vendor tool compatibility, of course.


I think a FileContent return type would make a great addition to the protocol. We are currently left either sequeezing multipart-mime into a text response or abusing the image response to pass non-image binary data, both of which is a hack.

**Reply by mheubi** [reply]
Actually, I think for now the safest way to return files from an MCP service is to abuse the ImageContent response by supplying it with a mime type that takes some creative license:

application/vnd.custom.binary; filename=xxx.ext; type=application/pdf

This essentially allows binary file transfer with filename and mime type attribution within the existing spec and it seems to pass through both SDKs without error.

**Reply by evalstate** [reply]
Which way round though? For Server->Client, you can just use an EmbeddedResource. For Client->Server you have roots (or the aforementioned "ImageContent" Sampling workaround)

**Reply by cpryland** [reply]
@evalstate Can you explain this further?

**Reply by mheubi** [reply]
server to client.

Currently the SDK restricts you to returning either TextContent or ImageContent. Due to the use of Pydantic and zod, anything not adhering to that is rejected. So the most adherent way to sneak other content through would be to use the "loophole" in ImageContent that it accepts any mime type, even non image ones. Doing so, ImageContent becomes sort of a more generic "BinaryContent".

@evalstate I realize that resources allow me to pull files from the server, but that just shifts the problem from "how does my tool return a named binary file" to "how does my tool return a reference to a resource". True, I can simply use text response, but then the client code needs knowledge of the tool, so it knows not to use the tool response literally but interpret it as a resource reference.

And this whole issue started on my side due to security considerations (i.e. passing data directly within the tool request communication and keeping it only transient, as opposed to placing it somewhere and making it available for any requester who happens to guess the right URI)

**Reply by cpryland** [reply]
What do the clients (models) do with these snuck-in imageContents?

**Reply by mheubi** [reply]
The MCP provides document conversion / text extraction. The result may be huge and must not be loaded directly into LLM context.
That is why it is passed back as a file, stored in a virtual filesystem inside the client.
The LLM then accesses it piecemeal by using additional tools.

The whole process is completely transparent:
1. The user drags a file into the client prompt.
2. It is stored in the virtual file system.
3. Once the user asks questions about it, the LLM tries to access it
4. The LLM realizes that it is non-text and invokes the MCP converter, which places a text file in the file system (as described above)
5. The LLM tries to access the converted file and realizes it is to big
6. The LLM then invokes additional tools (default ones, directly integrated into the client) to do the piecewise access.

**Reply by cpryland** [reply]
So you're depending on the client LLM to recognize all these facts and do the right thing?

Doesn't sound reliable...

**Reply by mheubi** [reply]
Well, yes and no….
The trick is to give the LLM guidance without having to actively manage control flow:
- the file reader tool, when instructed to read a read a file will yield a „file format is not text“ error on the pdf
- it will yield a „file is too big„ error if the file is Text but larger than, say, 50K

this in turn motivates the llm to invoke the file converter tool and the file splitter tool, respectively 

It actually works surprisingly reliable, if you’re willing to let the llm trigger those errors once rather than getting it immediately right

**Reply by cpryland** [reply]
In the case of producing a file, I wonder if the server couldn't output to the recently-created output, then return text/markdown with a link to the file either as a remote URL or a local file:/// URL?

At least then the user could have something to click on to view? Especially if this is an endpoint in the LLM tool usage chain?

**Reply by mheubi** [reply]
That's how it's normally done. However I'm bound by company policy that confidential information must not be physically stored in the AI subsystem. This is why I'm trying to pass the data directly to and from the tool and only keep it in session memory.

BTW here's what the LLM does with this toolset:
the query_file tool accesses chunked embeddings that have been asynchronously built as soon as the user uploaded the file.
(I agree, it's quite amazing how the LLM reliably figures all this out by itself)

{ContextElement: actor=ASSISTANT, content="How can I help you?"}
{FileData: actor:USER, content="[context][file id=69d767d0-M-Gruppe-FIBE-FER-2024_de.pdf name=M-Gruppe-FIBE-FER-2024_de.pdf type=application/pdf]"}
{ContextElement: actor=USER, content="**how many employees did migros have in 2023**"}
{**ToolInvocation**: id=call_1uEt2EceHOJAp3ZfD9u1kgJz, name=query_file, args={"id":"69d767d0-M-Gruppe-FIBE-FER-2024_de.pdf","**query**":"**number of employees in Migros in 2023**"}}
{ToolResult: success=false, id=call_1uEt2EceHOJAp3ZfD9u1kgJz, content=[Error] File ID: 69d767d0-M-Gruppe-FIBE-FER-2024_de.pdf - **File is not text**.}
{**ToolInvocation**: id=call_dmCQzDZwMzAgwciB0J2Vk788, name=**convert_to_text**, args={"file_id":"69d767d0-M-Gruppe-FIBE-FER-2024_de.pdf"}}
{ToolResult: success=true, id=call_dmCQzDZwMzAgwciB0J2Vk788, content=[File] File ID: 8140311b}
{**ToolInvocation**: id=call_GAKkDyNftVgkV2Tw65nppRve, name=**query_file**, args={"id":"8140311b","query":"**number of employees in Migros in 2023**"}}
{ToolResult: success=true, id=call_GAKkDyNftVgkV2Tw65nppRve, content=[Content] File ID: 8140311b - Migros Kompakt > **Kennzahlen** **5**...}
...
{ContextElement: actor=USER, content="**give me a summary of the whole document**"}
{**ToolInvocation**: id=call_O90EGlmx1kRIWIDBJiLoKaKY, name=**split_file**, args={"id":"8140311b"}}
{ToolResult: success=true, id=call_O90EGlmx1kRIWIDBJiLoKaKY, content=File ID: 8140311b split into 18 parts:
8140311b:0
8140311b:1
...}

**Comment by cpryland** [comment]
That would be great, but won't most MCP clients reject the mimeType `text/markdown`? Or are you saying you actually return that string for the LLM to pick up and vibe on later?

**Reply by mheubi** [reply]
The pdf extraction generates markdown, as it allows the llm to better analyze (or "vibe" ;-) the content (hierarchy of headings, for example).

The code above is just a proposal, currently there is not even a FileContent return type, only Text and Image. 
The ImageContent return type already implements mimeType, the proposal essentially only adds a file name and allows for the type to be literal 'file'.

**from SDK types.py:**
```
class ImageContent(BaseModel):
    """Image content for a message."""

    type: Literal["image"]
    data: str
    """The base64-encoded image data."""
    mimeType: str
    """
    The MIME type of the image. Different providers may support different
    image types.
    """
    annotations: Annotations | None = None
    model_config = ConfigDict(extra="allow")

```

**Comment by evalstate** [comment]
Hi, this PR is related : https://github.com/modelcontextprotocol/specification/pull/223 as a way to have Server's intelligently adapt content types based on what is tokenizable. I assume that everything which is text is "good" (although the Anthropic Messages API doesn't for example accept text/markdown as a document type - so juts send as text/plain). 

One other point on file transfer (not recommending this...) would be to have your Client respond to Sampling requests (e.g. parse the request and look for a resource name) and use the SamplingMessage to respond with your data.

**Reply by mheubi** [reply]
thx for pointing out the PR

**Comment by dang-1234** [comment]
I guess I have a similar issue of huge responses from an API and context size.
Is there a similar possible solve?

https://github.com/orgs/modelcontextprotocol/discussions/272

**Comment by ravi0613** [comment]
**How can I handle large file transfers between an MCP server and client without exceeding token limits in LLM calls?**

I'm using the Model Context Protocol (MCP) for a file processing pipeline where:
On the client side, I have an agent called File Processing Agent.This agent uses tools exposed by the MCP server.

The process:

I instruct the agent by providing a filename and URL.The agent invokes an MCP server tool that downloads and reads a large Excel file from the provided URL.
Let us assume the file has 4 sheets, each with around 4,000 rows and 12 columns (total ~16,000 rows).
The content is then converted to JSON and sent back to the client.
After receiving the file content, the agent is tasked to pass it to another tool that writes the content to a local file on the client.

The issue:

Due to the file’s size, the JSON content is massive.

The LLM (Azure OpenAI) throws a 400 Bad Request error stating:

Content length is more than 128,000 tokens, but got around 1.95 million tokens.

I’ve already tried:

1. Splitting the file into chunks and sending them sequentially.
2. Encoding the file in Base64.

But I still hit the token limit after processing 1–2 chunks, with Base64 still reaching around 1.12M tokens.

What I need help with:

Are there alternative strategies like:

1. Streaming?
3. External file references?
4.  Bypassing the LLM’s context for large data payloads?

What is the best approach to transfer large file content (like Excel data) using MCP, in a way that avoids hitting the token limit?

**Reply by tomelliot** [reply]
You're inherently limited by the protocol for this use case. If you have two separate tools, then the data has to pass through the LLM and do inference. Architecturally this is going to be bad practice - your data may get modified during the inference and you won't know about it. 

Some alternatives:
- write one single tool that does both tasks
- update both tools to use a filesystem - e.g. the first tool stores the file to disk, and the second tool reads it from disk again

**Reply by boylec** [reply]
Please see SEP-1708 which I have introduced and am taking ownership of. I *believe* it is related and may address the needs mentioned in this thread. Would love to hear your opinion(s).

https://github.com/modelcontextprotocol/modelcontextprotocol/issues/1708

**Reply by abhiram-ar** [reply]
@ravi0613 The core bottleneck here is that we’re asking an LLM to edit a very small portion of an Excel file, while LLMs operate under strict context window limits. Converting the entire Excel sheet into a JSON string and asking the model to modify it is both inefficient and unnecessary.

Instead, I’d propose introducing a local MCP server that exposes domain-specific tools for Excel manipulation. These tools would use the Excel API (or a suitable library) to programmatically modify the sheet.

For example:
1. Need to compute the average of column 2? → Expose a dedicated tool for that.
2. Need to insert or update a row? → Expose a tool for row-level mutations.

All the business logic and file-level operations would live inside the MCP server, while the LLM would only orchestrate which tool to call and with what parameters. This keeps the LLM out of large text processing, avoids context window issues, and results in a more scalable, deterministic, and maintainable system.


<img width="1258" height="703" alt="image" src="https://github.com/user-attachments/assets/2ed0de00-d5a9-405e-be17-afad75caaceb" />

**Reply by jiashuoz** [reply]
Would a CSV file work for your use case? Is Excel a must?

If CSVs work for your data analysis use cases, I'd love to invite you to become our pilot user at https://agentdrive.run/ (beta).

**Comment by fabb** [comment]
For my use case, i would need the MCP client to be able to upload any number of images/files to the MCP server, but i'm flexible if this happens before or after the MCP tool call.

**Reply by jiashuoz** [reply]
Would love to invite you to become our pilot user at https://agentdrive.run/ (beta).

**Comment by davispuh** [comment]
So what is the current approach to let LLMs work with large files without actually reading them?
Basically I want user to upload a file and then that file gets passed to MCP tool directly without LLM reading it. MCP tool would return small response which LLM then would interpret and reply to user.
How this could be accomplished?

**Reply by jiashuoz** [reply]
You can try https://agentdrive.run/ (beta). It provides a direct upload URL for your agent to POST a local file to the cloud. Would love to invite you as our pilot user. How large is your file?

**Reply by davispuh** [reply]
I don't care about your product. I'm looking for a way to implement this functionality for my app. How have you implemented it? It seems like typical way is LLM reads file and then puts content to upload tool but that's not what I want because that pollutes LLMs context.

**Reply by jiashuoz** [reply]
I’m just saying my product may help you solve this problem. It’s essentially a cloud storage system built for agents.

If you want to implement this yourself, you could get a direct upload URL and then have your agent initiate direct curl POST with a local file, that way llm never reads any file bytes.

happy to jump on a meeting to discuss more

**Comment by Erfouni** [comment]
Current status (July 2026): the published MCP `2025-11-25` protocol still has no interoperable **client → server file-upload primitive**. [`roots/list`](https://modelcontextprotocol.io/specification/2025-11-25/client/roots) returns `file://` URIs that describe filesystem scope; it does not transfer bytes or make a remote server able to read the client's disk. So supporting roots helps a local/shared-filesystem deployment, but it does not solve remote uploads.

The production pattern I would use today is to keep the data plane out of both the model context and the JSON-RPC argument text:

```text
user selects/attaches file
        ↓
client/host uploads bytes directly to trusted storage or the MCP server
        ↓
client/host injects an opaque file handle into tools/call
        ↓
server resolves the handle and processes the bytes
        ↓
tool returns only a small structured result (plus a resource link if needed)
```

For example, make the model-visible argument a logical reference:

```json
{
  file_id: file_01J...,
  name: report.xlsx,
  mime_type: application/vnd.openxmlformats-officedocument.spreadsheetml.sheet,
  size: 4281932,
  sha256: ...
}
```

The host should populate or replace `file_id` before dispatch; the model should never read, reproduce, or base64-encode the bytes. If you own only the server and not the host/client, there is currently no portable way to force this behavior: provide an authenticated upload UI/endpoint and pass its returned handle to the tool, or ship a local stdio companion that can read an explicitly approved path.

Security properties that matter for the handle/upload flow:

- bind the handle to the user, MCP server, session, and intended tool;
- use a short expiry and preferably one-time redemption;
- validate length, MIME type, digest, and filename server-side;
- use a server-minted/signed upload URL rather than letting the tool dereference an arbitrary model-supplied URL (SSRF/confused-deputy risk);
- canonicalize and allowlist paths for a local stdio server—roots are useful scope metadata, not a replacement for filesystem authorization;
- return a small result. For server → client artifacts, the current tool-result format already supports [`resource_link`](https://modelcontextprotocol.io/specification/2025-11-25/server/tools#resource-links), which is preferable to embedding a large payload.

There is active standards work for a portable version of this: [SEP-2631](https://github.com/modelcontextprotocol/modelcontextprotocol/pull/2631) proposes file objects plus `files/authorizeUpload` / `files/authorizeDownload`, keeping bytes out of JSON-RPC. It is still an open Draft, so feature-detect it and retain the handle/upload fallback rather than treating it as published MCP behavior.

**Reply by sbrady** [reply]
Have you got an example repository of this somewhere?

**Comment by fctr-id** [comment]
We ran into this exact context-bloat issue when handling large data payloads over MCP.

To avoid streaming raw files or large JSON dumps directly into the prompt context, we ended up offloading the processing:

- **Sampling & Server-Side Execution:** The tool returns only a few sample rows and metadata. The LLM then writes Python code to process and filter the data inside a server-side sandbox.
- **File Offloading:** Large result sets write out to local CSV files, so the LLM only gets back a file reference and summary.

If anyone wants to see how this is implemented in Python, the code is open-source here:
https://github.com/fctr-identity/fctr-okta-mcp-server

Feel free to take a look or grab ideas from it if it's helpful.

**Comment by SebTardif** [comment]
@nickwinder @davispuh

> So what is the current approach to let LLMs work with large files without actually reading them?

For a **local / shared-filesystem** deployment, do not upload the file at all. Pass a **path** (or a `file://` URI the server can open). The tool reads and writes on disk. The model only sees a short result: ok/fail, span, `error_kind`, maybe a small excerpt.

That is the opposite of “base64 in the tool call.” It also does not depend on `roots/list` transferring bytes. `roots/list` only advertises scope. A local server still needs its own path check (workspace containment) before it opens anything.

That is the model we use in [patchloom](https://github.com/patchloom/patchloom): MCP tools take workspace paths, never file blobs. Large JSON/YAML/markdown stays on disk. Structured replace/search/AST ops return JSON the agent can branch on, not the whole file. PathGuard rejects `../` and paths outside the workspace.

For a **remote** MCP server that cannot see the client disk, the July 2026 status in this thread still holds: there is no interoperable client-to-server file-upload primitive in the published spec. Keep the data plane out of JSON-RPC arguments:

1. Host uploads bytes to storage the server can read.
2. Tool argument is an opaque handle or URL.
3. Tool returns a summary or a handle, not 16k spreadsheet rows.

`roots/list` does not solve that remote case. Neither does asking the model to paste base64.

So: local editor / workspace tools should take paths. Remote processors should take handles. Do not put the file in the model context in either case.

**Comment by sealnet-team** [comment]
One distinction worth adding to the path/handle approach above: keeping file bytes out of context and keeping download credentials out of context are separate requirements. A URL can be small enough for the context window while still granting access to the entire file.

For a remote workflow, I would keep the mapping from an opaque file handle to its download URL/key inside trusted application code. The model passes the handle; the receiving tool resolves it after checking authorization, downloads the bytes, and returns only a status or output path. If both tools already share an approved filesystem, passing a local path is simpler and avoids an unnecessary upload.

Disclosure: I’m involved with SEAL. A concrete, narrower example is its MCP handoff mode: it shares an approved local file and returns a handle/receipt, while delivering the sharing link through a separate local channel. In a local SDK-client check of sealnet-mcp 0.2.4, the sharing URL was absent from the tool result and captured stderr; the received file matched the original SHA-256, and a further receive was rejected after revocation.

That is not a drop-in implementation of generic client-to-remote-server file inputs. It also does not isolate the link from an agent that can read the delivery file or clipboard—host permissions still matter. The check was not an independent security audit.

[SEAL MCP documentation](https://seal.net/docs/mcp?utm_source=github&utm_medium=community&utm_campaign=seal_growth_2026q4&utm_content=mcp_discussion_1197).

Drafted with AI assistance.
